### PR TITLE
Precharge implemented with accumulator voltage sensing

### PR DESCRIPTION
**Pull request by Eddy13013**
### Summary
Updated the precharge code to match our firmware structure. IO and App code are now abstracted from one another. Apologies for the long PR but this is a culmination of several months of testing and slow working. Made a large number of changes here to combine everything into one good PR.

### Changelist 
- Added underflow protection for Io_voltage sense
- Changed how we enter drive state vs. charge state. We don't want to check for OC when entering charge state. We expect OC.
- Created App code for the tractive system. There is a function to retrieve the TS voltage and the TS capacitance. Information about the TS is defined at the top of this file for easy access and characterization of the TS. This required making several changes to the main.c and BMS world code.
- Abstracted all IO code from the APP code.
- Added to the Precharge sequence app code. Here is more defined values for the precharge sequence. They can be retrieved with functions. The expected precharge time range can be retrieved from here as well as the precharge completion threshold.

### Testing Done

- Car was started with the previous code to this. With a threshold of 93% while pdm is connected to the wall. Didn't work with the 18650s installed. Believe that is a hardware issue. I've lowered that to our minimum of 90% which will hopefully work. If not, there will be a hardware solution.
- This current code has not been tested but should be functionally the same to the previous code. 

### Resolved Issues
- Abstracted Io code and App code

### Checklist
*Please change `[ ]` to `[x]` when you are ready.*
- [ ] I have read and followed the code conventions detailed in [README.md](../README.md) (*This will save time for both you and the reviewer!*).
- [ ] If this pull request is longer then **500** lines, I have provided *explicit* justification in the summary above explaining why I *cannot* break this up into multiple pull requests (*Small PR's are faster and less painful for everyone involved!*).

### CODE REVIEW

**Comment by pasitlaothamatas** [comment]
Make sure the IO code is abstracted from the App cide, and that there are no issues with build checks

File: boards/BMS/Src/App/App_PreChargeSequence.c [hunk 1]
```diff
@@ -2,6 +2,14 @@
 #include <stdlib.h>
 #include "App_PreChargeSequence.h"
 
+#define PRECHARGE_THRESHOLD 0.90f // EV.6.6.1: at least 90%
+#define PRECHARGE_TIME_LOWER_BOUND_PERCENTAGE 0.10f
+#define PRECHARGE_TIME_UPPER_BOUND_PERCENTAGE 1.50f
+#define PRECHARGE_RESISTANCE_OHMS (1000U)
+// COEFFICIENT_TO_THRESHOLD is the result of solving the exponential (for time
+// to threshold) divided by RC (i.e. 4 would correspond to 98%)
+#define COEFFICIENT_TO_THRESHOLD 2.30f
```
Comment (pasitlaothamatas) [inline boards/BMS/Src/App/App_PreChargeSequence.c:11]: im kinda lost here what is this value supposed to be?

File: boards/BMS/Inc/App/App_Accumulator.h [hunk 2]
```diff
@@ -43,6 +65,17 @@ struct Accumulator *App_Accumulator_Create(
...
+/**
+ * Get the pack voltage in V
+ * @param accumulator The accumulator to get the pack voltage
+ * @return The accumulator pack voltage in V
+ */
+float App_Accumulator_GetPackVoltage(struct Accumulator *accumulator);
+
+bool App_Accumulator_IsMaxVoltageInRange(struct Accumulator *accumulator);
+
+bool App_Accumulator_IsMinVoltageInRange(struct Accumulator *accumulator);
```
Comment (pasitlaothamatas) [inline boards/BMS/Inc/App/App_Accumulator.h:77]: can you add documentation for these fxns

File: boards/BMS/Inc/App/App_PreChargeSequence.h [hunk 3]
```diff
@@ -37,3 +38,37 @@ void App_PreChargeSequence_Enable(
...
+ * The pre-charge sequence should be quicker than this time.
+ * @param pre_charge_sequence The pre-charge sequence to estimate for
+ * @param tractive_system_capacitance The capacitance of the connected tractive
+ * system
+ * @return precharge_upper_time_bound The upper time bound for the pre-charge
+ * sequence
+ */
+uint32_t App_PreChargeSequence_GetLowerTimeBound_MS(
+    const struct PreChargeSequence *const pre_charge_sequence,
+    float                           tractive_system_capacitance);
```
Comment (pasitlaothamatas) [inline boards/BMS/Inc/App/App_PreChargeSequence.h:74]: we should put these #defines in a header file somewhere, dont need to use getter functions

File: boards/BMS/Inc/App/App_TractiveSystem.h [hunk 4]
```diff
@@ -0,0 +1,36 @@
+#pragma once
+
+#include "App_SharedBinaryStatus.h"
```
Comment (pasitlaothamatas) [inline boards/BMS/Inc/App/App_TractiveSystem.h:3]: i dont think his header is needed any longer?

File: boards/BMS/Src/Io/Io_VoltageSense.c [hunk 5]
```diff
@@ -30,6 +31,14 @@ float Io_VoltageSense_GetTractiveSystemVoltage(float adc_voltage)
     // TS_Voltage = ---------------------------------- * Measurement Error %
     //                Voltage Ratio x Amplifier Gain
 
-    return adc_voltage * R_ERROR_COMPENSATION /
-           (TS_VOLTAGE_DIV * AMPLIFIER_GAIN);
+    if (adc_voltage < 0.0f)
+    {
+        return NAN;
+    }
+
```
Comment (pasitlaothamatas) [inline boards/BMS/Src/Io/Io_VoltageSense.c:38]:
```suggestion
```

**Comment by pasitlaothamatas** [comment]
can you remove eeprom config stuff for the DCM? its out of scope

File: boards/BMS/Src/App/App_TractiveSystem.c [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+float App_TractiveSystem_GetSystemVoltage(const
+    struct TractiveSystem *const tractive_system)
+{
+    return tractive_system->get_system_voltage(tractive_system->get_adc_voltage());
+}
+
+float App_TractiveSystem_GetSystemCapacitance(void)
+{
+    return INVERTER_CAPACITANCE_F * NUM_INVERTERS;
+}
```
Comment (pasitlaothamatas) [inline boards/BMS/Src/App/App_TractiveSystem.c:43]: as mentioned previously this should just be a pound define

File: boards/BMS/Src/App/states/App_AllStates.c [hunk 7]
```diff
@@ -17,6 +17,7 @@ void App_AllStatesRunOnTick1Hz(struct StateMachine *const state_machine)
     App_CanTx_SetPeriodicSignal_IS_CONNECTED(can_tx, charger_is_connected);
 }
 
+#include "Io_Airs.h"
```
Comment (pasitlaothamatas) [inline boards/BMS/Src/App/states/App_AllStates.c:20]: make sure the includes are all done at the top of the file

Comment (pasitlaothamatas) [reply boards/BMS/Src/App/states/App_AllStates.c:20]: check other files too

File: boards/BMS/Src/App/states/App_DriveState.c [hunk 8]
```diff
@@ -17,6 +17,8 @@ static void DriveStateRunOnTick1Hz(struct StateMachine *const state_machine)
     App_AllStatesRunOnTick1Hz(state_machine);
 }
 
+#include "Io_Airs.h"
+#include "states/App_FaultState.h"
```
Comment (pasitlaothamatas) [inline boards/BMS/Src/App/states/App_DriveState.c:21]: move declaration to top of the file, remove io_airs.h

File: boards/BMS/Src/Io/Io_Airs.c [hunk 9]
```diff
@@ -1,35 +1,51 @@
...
 bool Io_Airs_IsAirPositiveClosed(void)
 {
-    return HAL_GPIO_ReadPin(AIR_TOTAL_ISENSE_GPIO_Port, AIR_TOTAL_ISENSE_Pin) &&
-           HAL_GPIO_ReadPin(AIR_POWER_STATUS_GPIO_Port, AIR_POWER_STATUS_Pin);
+    // const float air_loop_current =
+    //    (float)Io_Adc_GetAdc2Channel4Voltage() * AIR_TOTAL_ISENSE_V_TO_A;
+    // return HAL_GPIO_ReadPin(AIR_POWER_STATUS_GPIO_Port, AIR_POWER_STATUS_Pin)
+    // &&
+    //       (air_loop_current > AIR_POSITIVE_CLOSED_THRESH_A);
+
```
Comment (pasitlaothamatas) [inline boards/BMS/Src/Io/Io_Airs.c:32]: i wanted to determine what this current would be. we should be able to get rid of is_air_positive_closed when we merge this in

File: boards/BMS/Src/App/App_TractiveSystem.c [hunk 10]
```diff
@@ -0,0 +1,43 @@
+#include "App_TractiveSystem.h"
+#include <stdlib.h>
+#include <assert.h>
+#include "App_SharedMacros.h"
```
Comment (pasitlaothamatas) [inline boards/BMS/Src/App/App_TractiveSystem.c:4]: convention

```suggestion
#include <stdlib.h>
#include <assert.h>
#include "App_TractiveSystem.h"
#include "App_SharedMacros.h"
```

**Comment by pasitlaothamatas** [comment]
You need to rebuild before you push to run clang-format on the code. Check and see if tests are still failing after that

**Comment by pasitlaothamatas** [comment]
Also still notice some IO code in the App code, that needs some cleaning up

**Comment by pasitlaothamatas** [comment]
App_PrechargeSequence should be deleted. There is already the precharge state to handle precharge, The constants should be moved to a .h file which means that we can get rid of the getter functions. Following convention we may need something like a `App_PrechargeRelay` that we can use in the state machine

File: boards/BMS/Src/App/states/App_PreChargeState.c [hunk 11]
```diff
@@ -20,11 +35,100 @@ static void
...
+        App_Accumulator_GetPackVoltage(accumulator) *
+        App_PreChargeSequence_GetThreshold_Percent(pre_charge_sequence);
+
+    // Time since precharge started
+    uint32_t elapsed_time =
+        App_SharedClock_GetCurrentTimeInMilliseconds(clock) -
+        App_SharedClock_GetPreviousTimeInMilliseconds(clock);
+
+    // CHARGE MODE
+    if (App_Charger_IsConnected(charger))
```
Comment (pasitlaothamatas) [inline boards/BMS/Src/App/states/App_PreChargeState.c:71]: should avoid repeating code here. should do precharge first regardless of whether the charger is connected then decide which state to go to next based on whether precharge has been completed

File: boards/BMS/Src/App/states/App_PreChargeState.c [hunk 12]
```diff
@@ -20,11 +35,100 @@ static void
...
+
+    // lower and upper bounds on expected charge time. Used to recognize faults
+    const uint32_t lower_precharge_time =
+        App_PreChargeSequence_GetLowerTimeBound_MS(
+            pre_charge_sequence,
+        App_TractiveSystem_GetSystemCapacitance());
+    const uint32_t upper_precharge_time =
+        App_PreChargeSequence_GetUpperTimeBound_MS(
+            pre_charge_sequence,
+            App_TractiveSystem_GetSystemCapacitance());
```
Comment (pasitlaothamatas) [inline boards/BMS/Src/App/states/App_PreChargeState.c:57]: yeah as previously mentioned we can avoid using getter fxns and use pound defines

File: boards/BMS/Inc/App/App_Accumulator.h (hunk 2 above)
Comment (Eddy13013) [reply boards/BMS/Inc/App/App_Accumulator.h:77]: added

File: boards/BMS/Inc/App/App_TractiveSystem.h (hunk 4 above)
Comment (Eddy13013) [reply boards/BMS/Inc/App/App_TractiveSystem.h:3]: deleted

File: boards/BMS/Src/Io/Io_VoltageSense.c (hunk 5 above)
Comment (Eddy13013) [reply boards/BMS/Src/Io/Io_VoltageSense.c:38]: deleted

File: boards/BMS/Src/App/App_PreChargeSequence.c [hunk 13]
```diff
@@ -41,3 +49,25 @@ void App_PreChargeSequence_Disable(
...
+    return (uint32_t)(COEFFICIENT_TO_THRESHOLD * PRECHARGE_RESISTANCE_OHMS *
+           tractive_system_capacitance * (float)PRECHARGE_TIME_UPPER_BOUND_PERCENTAGE);
+}
+
+uint32_t App_PreChargeSequence_GetLowerTimeBound_MS(
+    const struct PreChargeSequence *const pre_charge_sequence,
+    float                                 tractive_system_capacitance)
+{
+    return (uint32_t)(COEFFICIENT_TO_THRESHOLD * PRECHARGE_RESISTANCE_OHMS *
+           tractive_system_capacitance * (float)PRECHARGE_TIME_UPPER_BOUND_PERCENTAGE);
```
Comment (timbitss) [inline boards/BMS/Src/App/App_PreChargeSequence.c:72]: should be lower_bound_percentage

**Pull request closed without merging**